src/torchcodec/decoders/_core/VideoDecoder.cpp

```diff
@@ -187,18 +187,34 @@ VideoDecoder::VideoStreamDecoderOptions::VideoStreamDecoderOptions(
   }
 }
 
-VideoDecoder::BatchDecodedOutput::BatchDecodedOutput(
-    int64_t numFrames,
-    const VideoStreamDecoderOptions& options,
-    const StreamMetadata& metadata)
-    : frames(torch::empty(
-          {numFrames,
-           options.height.value_or(*metadata.height),
-           options.width.value_or(*metadata.width),
-           3},
-          at::TensorOptions(options.device).dtype(torch::kUInt8))),
-      ptsSeconds(torch::empty({numFrames}, {torch::kFloat64})),
-      durationSeconds(torch::empty({numFrames}, {torch::kFloat64})) {}
+torch::Tensor VideoDecoder::allocateEmptyHWCTensorForStream(
+    int streamIndex,
+    std::optional<int> numFrames) {
+  auto metadata = containerMetadata_.streams[streamIndex];
+  auto options = streams_[streamIndex].options;
+  auto height = options.height.value_or(*metadata.height);
+  auto width = options.width.value_or(*metadata.width);
+
+  auto tensorOptions = torch::TensorOptions()
+                           .dtype(torch::kUInt8)
+                           .layout(torch::kStrided)
+                           .device(options.device.type());
+  if (numFrames.has_value()) {
+    return torch::empty({numFrames.value(), height, width, 3}, tensorOptions);
+  } else {
+    return torch::empty({height, width, 3}, tensorOptions);
+  }
+}
+
+VideoDecoder::BatchDecodedOutput VideoDecoder::allocateBatchDecodedOutput(
```

Comment:
The previous `BatchDecodedOutput` constructor wasn't a member of `VideoDecoder`, and so I got compilation errors because we're calling `allocateEmptyHWCTensorForStream` which itself is a member method (because it accesses attributes blahblahblah).

So I had to make the "constructor" a member itself.

Reply:
I understand that but by making it a member we are letting it access a lot more information than it strictly needs. It only needs height and width, and not all the other stuff in VideoDecoder. What's worse is it can accidentally change the state of the VideoDecoder because it's not `const`

I think it may be better to pass in the height and width after looking it up in the caller.

Reply:
A general principle I like to apply is that classes should know how to create correct objects, and ideally, constructed objects should be valid. That means classes should have constructors, and the construtor is responsible for making the object valid. So I think we should keep the `BatchDecodedOutput` explicit constructor. If `VideoDecoder` then needs to construct `BatchDecodedOutput` objects using some of its own internal state, then `VideoDecoder` should figure out what's relevant to `BatchDecodedOutput` and pass it in.

We may need a `VideoDecoder::allocatedBatchDecodedOutput()` method, but if we do, it should call `BatchDecodedOutput`'s explicit constructor. What this could potentially look like:
```
VideoDecoder::BatchDecodedOutput::BatchDecodedOutput(
  int64_t numFrames,
  torch::Tensor frames)
  : frames(frames),
  ptsSeconds(torch::empty({numFrames}, {torch::kFloat64})),
  durationSeconds(torch::empty({numFrames}, {torch::kFloat64})) {}

VideoDecoder::BatchDecodedOutput VideoDecoder::allocateBatchDecodedOutput(
  int streamIndex,
  int64_t numFrames) {
  return BatchDecodedOutput(
    numFrames,
    allocateEmptyHWCTensorForStream(streamIndex, numFrames)
  );
}
```
We can take advantage of tensors have reasonable move-like semantics.